### PR TITLE
Add support of `M/Illustration` metadata in zim file.

### PR DESCRIPTION
**Pull request by mgautierfr**
Remove favicon API and replace it by illustration methods.

We have decided with @kelson42 to not implement resolution fallback for this PR. Has I had already implemented a small fallback algorithm, I've keep it commented as reference.
It is not to be reviewed as the exact algorithm will probably change.

We need new zim file with the correct illustration entry. I've made a prerealase of zim-testing-suite with them (https://github.com/openzim/zim-testing-suite/releases/tag/v0.3-alpha). v0.3-alpha is made using this branch and https://github.com/openzim/zim-tools/pull/241

codecov is a bit under 90% has we don't test the fallback if there is no illustration/favicon at all in the zim file.

### CODE REVIEW

File: include/zim/archive.h
Comment (veloman-yunkan) [inline include/zim/archive.h:152]: What does this return if the illustration of the exact requested size doesn't exist in the archive? The behaviour for that case must be documented.

File: include/zim/archive.h
Comment (veloman-yunkan) [inline include/zim/archive.h:265]: `archive.hasIllustration(size)` can be easily confused for `archive.hasIllustration(size, size)` whereas it actually stands for `archive.hasIllustration(size, 48)`. In such cases it is better to have two functions:

```c++
    bool hasIllustration(unsigned int width, unsigned int height) const;
    bool hasIllustration(unsigned int size=48) const { return hasIllustration(size, size); }
```

File: include/zim/writer/creator.h
Comment (veloman-yunkan) [inline include/zim/writer/creator.h:147]: If content is required to be a PNG image, asking for its dimensions in the interface signature is redundant. A safer interface is to validate the content independently and extract its size automatically. Of course doing this will add to libzim a new dependency (libpng), however it will remove the burden from future clients of this interface.

File: include/zim/writer/creator.h
Comment (veloman-yunkan) [inline include/zim/writer/creator.h:156]: Same here

File: src/archive.cpp
Comment (veloman-yunkan) [inline src/archive.cpp:115]: So `getIllustrationItem(width, height)` can silently return an item of a different size. Maybe the return type must be adjusted so that the caller knows what (s)he got? Of course that can be guessed from the item's path, but having it in the interface is more fool-proof.

File: src/archive.cpp
Comment (veloman-yunkan) [inline src/archive.cpp:126]: On old ZIM archives this function will mostly cheat. What about a different interface that enumerates available illustrations?

File: test/creator.cpp
Comment (veloman-yunkan) [inline test/creator.cpp:201]: This test nicely demonstrates that `addIllustration()` is not fool-proof. Sure enough, there will be users running against this imperfection, we may have extra issues in the bug-tracker and ZIM archives with invalid illustration content.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
IMHO, the biggest problem with this PR is that the proposed interface is not fool-proof. I don't think that [worse-is-better](https://en.wikipedia.org/wiki/Worse_is_better) approach is justified in this case.

**Comment by mgautierfr** [comment]
Most of your comments are about make the libzim smart (parse content and deduce information about it).
`libzim` tends to do the opposite : Do not infer anything about the content :
- We don't parse the content to know its mimetype, we let the user provide it.
- We've stopped parsing html/css content to change the url.
- We add a api to provide what content to index (we still have a default html parser but it is here mainly for compatibility reason, the user should have its own parser)

I would like to keep it this way.
If there is a conflict between that and the fact that we enforce a PNG, I would rather drop the PNG enforcing.

---

On the API side, this is not so easy.
One thing that we haven't implement (and think about) is the pixel density.
A 48x48 illustration on a high density screen is a 96x96 pixels image.
But a "real" 96x96 illustration on a standard density screen is also a 96x96 pixels image. But the two images may be different
As free desktop says : "For instance, an icon with a directory size of 48 but scale 2x would be 96x96 pixels, but designed to have the same level of detail as a 48x48 icon at scale 1x. This can be used on high density displays where a 48x48 icon would be too small (or ugly upscaled) and a normal 96x96 icon would have a lot of detail that is hard to see." (https://specifications.freedesktop.org/icon-theme-spec/icon-theme-spec-latest.html#definitions)

My first implementation had only one size (assuming height==width). But as @kelson42 said me, illustration may be more than a simple icons, we may use wider illustration to use as banner. So I split size in height and width.

What I want here is :
- A spec good enough to be extended (or allow nice fallback) on the long term.
- A api good usable (stable) on short/middle term.

Rethinking about this, I think we should reduce the scope to :
- Illustration is a replacement of 1x scale, square icon previously called favicon (for now). The api should be made this may (only one size, fallback to favicon)
- On the format "M/Illustration_*" is also a replacement for square favicon. We may add a scale information in the path to be future proof.
- IF we want banner/background/whatever.. , we will add a new metadata called "M/banner_**"

@kelson42, @veloman-yunkan, @rgaudin seems good to you ?

**Comment by kelson42** [comment]
@mgautierfr @veloman-yunkan Would be good to move on this one. In general I’m not in favour of making libzim smart. I’m against the idea of having a libpng dependence in libzim. I would propose to transfer improvements in dedicated tickets so the can be discussed separatly and for the less urgent, later.

**Comment by kelson42** [comment]
@mgautierfr It is important we have a primitive to deliver the list od illustrations to be able to include them in OPDS stream.

**Comment by mgautierfr** [comment]
Following my proposal :

- Illustration is stored using the path `M/Illustration_<width>x<height>@<scale>`. This format is mostly what is used by freedesktop and web community.
  It will allow use to be future proof in the zim format.
- The API (both creator and archive) is now for a square icon (width==height) "normal" density (scale==1). We will change this when needed.
- There is now a method to get the list of available size (`getIllustrationSizes`).
- We consider the old `-/favicon` as a 48x48@1 illustration.

**Comment by kelson42** [comment]
I have rebased this branch

**Comment by kelson42** [comment]
@mgautierfr We need as well the corresponding PR in kiwix/libkiwix, otherwise this will break its compilation.

File: src/archive.cpp
Comment (veloman-yunkan) [inline src/archive.cpp:119]: This would accept `"Illustration_64x64@1.5"` too (and, as far as I know, non-integer scale values are allowed).

Using good old `sscanf()` is justified here:

```c++
unsigned int parseIllustrationPathToSize(const std::string& s) {
    int n;
    unsigned int w, h;
    if ( sscanf(s.c_str(), "Illustration_%ux%u@1%n)", &w, &h, &n) == 2
         && n == s.size() && w == h) {
        return w;
    }

    throw std::runtime_error("");
}
```

The only problem with this version is that it ignores any whitespace before the size (e.g. `"Illustration_    64x 64"` is happily accepted). A similar issue is also present in your version, however only if both width and height values are preceded with the same amount of whitespace (and it can be fixed easier in your implementation).

File: src/archive.cpp
Comment (veloman-yunkan) [inline src/archive.cpp:157]: Why don't you merge these two lines into one?

File: src/archive.cpp
Comment (veloman-yunkan) [inline src/archive.cpp:150]: A `for` loop without a condition will make this code slightly more readable. Using `Archive::findByPath()` that returns an `EntryRange` would be even better at the cost of slight performance loss (which, IMHO, is acceptable here, since `Archive::getIllustrationSizes()` is hardly going to be bottleneck).

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
A few suggestions to improve the code.

File: include/zim/archive.h
Comment (mgautierfr) [reply include/zim/archive.h]: Done

File: src/archive.cpp
Comment (mgautierfr) [reply src/archive.cpp]: Done, based on your implementation (in several commits).

File: src/archive.cpp
Comment (mgautierfr) [reply src/archive.cpp]: Done

File: src/archive.cpp
Comment (mgautierfr) [reply src/archive.cpp]: I've used a for loop.

The problem of `findByPath` is that it searches in `C` namespace only and we are searching in `M`.

File: src/archive.cpp
Comment (veloman-yunkan) [inline src/archive.cpp]: Nice improvement!

I figured out one more potential issue - since %n doesn't count against conversions returned as a result of `scanf()`, inputs like `"Illustration_NNxNN"` followed by anything other than `"@1"` may also be randomly accepted (if the uninitialized value of `nEnd` happens to be equal to 2). The fix is easy - initialize `nEnd` to 0.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
One little fix and we are good to merge. You are welcome to rebase/fixup so that you can immediately merge next.

**Comment by mgautierfr** [comment]
Done and almost ready to merge.

This PR use the alpha release of zim-testing-suite data. But the data contains change for this PR but also for #575.

Before removing the last commit and merging this PR, we need to do a real release of zim-testing-suite and for that we need to validate #575 (at least the zim creation part). Once #575 is ok, it will do the proper release/merge.

File: src/archive.cpp
Comment (mgautierfr) [reply src/archive.cpp]: Done

**Comment by kelson42** [comment]
@mgautierfr Does this PR needs a new review of @veloman-yunkan ?

**Comment by mgautierfr** [comment]
No, we already agree on it.
But we can't merge it right now, we must agree on #575 first.

**Pull request merged**
